### PR TITLE
DMA Debugging and Instrumenting

### PR DESCRIPTION
**Pull request by sjkelly**
This is an attempt to understand why DMA is erroring while keeping things simple with a single threaded example. The root cause as far as I can tell is that we cannot keep up with the buffers. One approach I would like to evaluate is sign extension on the FPGA (for processing savings). 
I am also not convinced our handling of overflows in the Soapy driver is correct. I notice that we are over-eagerly evacuating buffers when the system may catch up.
Of note we need to keep up 116MB/s processing speed. Failures occur at any lower rate than this.

### CODE REVIEW

File: software/litepcie-kernel-module/config.h [hunk 1]
```diff
@@ -55,7 +55,7 @@
 #define DMA_CHANNEL_COUNT DMA_CHANNELS
 #define DMA_BUFFER_PER_IRQ 32
 #define DMA_BUFFER_COUNT 256
-#define DMA_BUFFER_SIZE 8192
+#define DMA_BUFFER_SIZE 65536
```
Comment (sjkelly) [inline software/litepcie-kernel-module/config.h:58]: @maleadt this seems okay to do, correct?
This puts it right on the page limit for the GPU, but it doesn't seem to be a problem.

**Comment by sjkelly** [comment]
Here are some instrumenting results with a larger MTU by https://github.com/JuliaComputing/xtrx_julia/pull/72/commits/8a5cec6d173f5884dd49b3bdcdbebcfa6154756a:
```
[ Info: Receiving data using CPU with LFSR...
 ────────────────────────────────────────────────────────────────────────────
                                    Time                    Allocations      
                           ───────────────────────   ────────────────────────
     Tot / % measured:          2.72s /  98.5%            712KiB /  65.8%    

 Section           ncalls     time    %tot     avg     alloc    %tot      avg
 ────────────────────────────────────────────────────────────────────────────
 data validation    5.00k    1.53s   57.2%   306μs    234KiB   50.0%    48.0B
 acquire            5.00k    1.15s   42.8%   229μs    234KiB   50.0%    48.0B
 release            5.00k   1.15ms    0.0%   230ns     0.00B    0.0%    0.00B
 ────────────────────────────────────────────────────────────────────────────
[ Info: Data rate: 116.480 MiB/s
[ Info: Overflow events: 0

[ Info: Using CPU
[ Info: Receiving data using CPU with pattern...
 ────────────────────────────────────────────────────────────────────────────
                                    Time                    Allocations      
                           ───────────────────────   ────────────────────────
     Tot / % measured:          2.72s /  98.5%            738MiB / 100.0%    

 Section           ncalls     time    %tot     avg     alloc    %tot      avg
 ────────────────────────────────────────────────────────────────────────────
 data validation    5.00k    1.78s   66.5%   356μs    738MiB  100.0%   151KiB
 acquire            5.00k    897ms   33.5%   179μs    234KiB    0.0%    48.0B
 release            5.00k   1.42ms    0.1%   283ns     0.00B    0.0%    0.00B
 ────────────────────────────────────────────────────────────────────────────
[ Info: Data rate: 116.464 MiB/s
[ Info: Overflow events: 0


[ Info: Receiving data using GPU with pattern...
 ────────────────────────────────────────────────────────────────────────────
                                    Time                    Allocations      
                           ───────────────────────   ────────────────────────
     Tot / % measured:          2.72s /  98.5%            740MiB / 100.0%    

 Section           ncalls     time    %tot     avg     alloc    %tot      avg
 ────────────────────────────────────────────────────────────────────────────
 data validation    5.00k    1.87s   69.8%   374μs    739MiB  100.0%   151KiB
 acquire            5.00k    809ms   30.2%   162μs    234KiB    0.0%    48.0B
 release            5.00k   1.87ms    0.1%   375ns     0.00B    0.0%    0.00B
 ────────────────────────────────────────────────────────────────────────────
```

It seems that `acquire` is quite slow compared to `release`. Bumping the MTU should bring down the time spent acquiring buffers.
I also restored the flush-on-overflow behavior I originally removed in this PR. Even under these conditions, being sufficiently fast in the loops is fine. So my conclusion is we just need to be faster :)

**Comment by maleadt** [comment]
> I am also not convinced our handling of overflows in the Soapy driver is correct. I notice that we are over-eagerly evacuating buffers when the system may catch up.

Could you elaborate? You mean: https://github.com/JuliaComputing/xtrx_julia/blob/ab3678c57b34ef8dde7c82784422765586acccdd/software/soapysdr-xtrx/Streaming.cpp#L251-L256

Yeah that does drain some buffers too many, but if we do it more precisely we risk running into a new overflow right after again.

File: software/litepcie-kernel-module/config.h (hunk 1 above)
Comment (maleadt) [reply software/litepcie-kernel-module/config.h:58]: It's been a while since I wrote the page handling logic, but I think this is OK.

**Comment by sjkelly** [comment]
Yes. I noticed better behavior when changing the condition [here](https://github.com/JuliaComputing/xtrx_julia/blob/ab3678c57b34ef8dde7c82784422765586acccdd/software/soapysdr-xtrx/Streaming.cpp#L249-L250) to be `>32` rather than `((int64_t)_dma_mmap_info.dma_rx_buf_count / 2)`. Maybe `((int64_t)_dma_mmap_info.dma_rx_buf_count / 4)` would be a better value?

**Comment by maleadt** [comment]
Why? That would call it an overflow if the hardware gets ahead of the software by only 32 buffers. It may be that it better matches a Julia-based consumer (where we want ample headroom for an occasional pause), but in general that seems like a bad solution because we'd be running into overflows much more frequently. In principle the check should just be `if ((hw_count - sw_count) > dma_rx_buf_count))`, but the problem with that is that the detection might happen too late in which case the DMA has already overwritten the buffer that you're currently still processing.

**Comment by staticfloat** [comment]
What constrains us from just increasing the number of buffers from 256 to something much larger like 8K? With 8K buffers, at 1024 samples per buffer, if we operate at 40MHz sample rate (which we hope to be able to do) and we constrain ourselves to never have more than 1/2 of the buffers full, we give ourselves ~100ms to do things like GC pause, hiccup, etc...  That's more than enough, even in the worst case, so I think it's worth just trying that if it's easy.

Increasing the buffer size itself from 1K -> 64K has a similar effect, the main difference being that that increases the minimum latency of the system, whereas increasing the maximum buffers does not (since we should be able to control where in the circular buffers we are writing to/reading from; e.g. we can ["race the beam"](https://web.archive.org/web/20200114112848/http://blogs.valvesoftware.com/abrash/latency-the-sine-qua-non-of-ar-and-vr/) and write to/read from some variable distance ahead of the FPGA's read pointer/behind the FPGA's write pointer.  We can tune that value to determine the true latency overhead of the DMA system.

The minimum buffer size we want to roll with should be measured by what we think the minimum latency it is reasonable to support is.  Since we're generally optimizing for MIMO here, we're not going to bother supporting anything higher than 70MHz sampling rate, as that is what we can achieve with 8 channels (half the XTRX's) sampling full-bore and saturating our PCI-e bus.  I'm assuming the absolute highest-sampling-rate scenario here, where we're not actually transmitting back half of our channels, we're using them for synchronization with FPGA support.  Yes, that's not something we're going to do this year, but let's think ahead:

If PCI-e overhead is roughly 0.8 us (taken from [slide 12 of this presentation](https://gianniantichi.github.io/files/presentations/pciebench-sigcomm.pdf)), we know that our buffer sizes should be significantly larger than that (let's say at least 100x), to avoid spending too much time on overhead; that means that we should set our minimum buffer size to something like 5600 samples.

**Comment by sjkelly** [comment]
I don't believe there is a hard constraint, though I recall increasing the `table_depth` caused issues. I will retest now with these stack of changes and see if the improved validation helped.

**Comment by sjkelly** [comment]
Maybe there is. Placement fails with:
```
INFO: [Vivado_Tcl 4-199] Please refer to the DRC report (report_drc) for more information.
Running DRC as a precondition to command place_design
INFO: [DRC 23-27] Running DRC with 8 threads
ERROR: [DRC UTLZ-1] Resource utilization: LUT as Distributed RAM over-utilized in Top Level Design (This design requires more LUT as Distributed RAM cells than are available in the target device. This design requires 31904 of such cell types but only 9600 compatible sites are available in the target device. Please analyze your synthesis results and constraints to ensure the design is mapped to Xilinx primitives as expected. If so, please consider targeting a larger device. Please set tcl parameter "drc.disableLUTOverUtilError" to 1 to change this error to warning.)
ERROR: [DRC UTLZ-1] Resource utilization: LUT as Memory over-utilized in Top Level Design (This design requires more LUT as Memory cells than are available in the target device. This design requires 31917 of such cell types but only 9600 compatible sites are available in the target device. Please analyze your synthesis results and constraints to ensure the design is mapped to Xilinx primitives as expected. If so, please consider targeting a larger device. Please set tcl parameter "drc.disableLUTOverUtilError" to 1 to change this error to warning.)
ERROR: [DRC UTLZ-1] Resource utilization: RAMD64E over-utilized in Top Level Design (This design requires more RAMD64E cells than are available in the target device. This design requires 31744 of such cell types but only 9600 compatible sites are available in the target device. Please analyze your synthesis results and constraints to ensure the design is mapped to Xilinx primitives as expected. If so, please consider targeting a larger device.)
ERROR: [DRC UTLZ-1] Resource utilization: Slice LUTs over-utilized in Top Level Design (This design requires more Slice LUTs cells than are available in the target device. This design requires 48026 of such cell types but only 32600 compatible sites are available in the target device. Please analyze your synthesis results and constraints to ensure the design is mapped to Xilinx primitives as expected. If so, please consider targeting a larger device. Please set tcl parameter "drc.disableLUTOverUtilError" to 1 to change this error to warning.)
WARNING: [DRC RPBF-3] IO port buffering is incomplete: Device port i2c0_scl expects both input and output buffering but the buffers are incomplete.
WARNING: [DRC RPBF-3] IO port buffering is incomplete: Device port i2c1_scl expects both input and output buffering but the buffers are incomplete.
INFO: [Vivado_Tcl 4-198] DRC finished with 4 Errors, 2 Warnings
INFO: [Vivado_Tcl 4-199] Please refer to the DRC report (report_drc) for more information.
ERROR: [Vivado_Tcl 4-23] Error(s) found during DRC. Placer not run.
INFO: [Common 17-83] Releasing license: Implementation
8 Infos, 2 Warnings, 0 Critical Warnings and 5 Errors encountered.
place_design failed
place_design: Time (s): cpu = 00:00:24 ; elapsed = 00:00:10 . Memory (MB): peak = 4004.664 ; gain = 0.000 ; free physical = 30931 ; free virtual = 53601
ERROR: [Common 17-39] 'place_design' failed due to earlier errors.

    while executing
```

**Comment by staticfloat** [comment]
Yeah, that sounds like we've run out of FPGA resources.  I remember you wondering if we were explicitly placing our DMA buffer area on BRAM, or if we're placing onto generic LUTs, did you ever find that out?  I took a brief look into `litepcie` to see if I could figure the answer out, and it looks to me like we use [generic `stream.SyncFIFO()` values](https://github.com/enjoy-digital/litepcie/blob/3c4d33ba64ad59307fcc7bfda75dc1d45a41b334/litepcie/frontend/dma.py#L101), which I think are probably lowered to generic LUT resources.  Looking at the [Artix-7 overview](https://docs.xilinx.com/v/u/en-US/ds180_7Series_Overview), we can synthesize ~600KB of RAM encoded in generic FPGA resources.

I'm assuming what you did is you tried to set buffer size to 8K samples (which corresponds to the buffer size in bytes of 64K in this PR) and number of buffers to 8K.  That would require 1MB of distributed RAM (512KB for reader and writer), which is definitely over our memory budget.  Our Artix-7 has over 1MB of BRAM available in it, so it's probably worthwhile for us to have Florent look into mapping the DMA reader/writer tables onto the BRAM in the long-term.

In the short-term, if we cut the table depth down by a factor of 8, we should have plenty of space, and I think that's a worthwhile change to make.

File: software/scripts/test_pattern.jl [hunk 2]
```diff
@@ -69,19 +73,27 @@ function dma_test(dev_args;use_gpu=false, lfsr_mode=false)
 
         test_mode = lfsr_mode ? "LFSR" : "pattern"
 
+        error_count = 0
+        errored_buffer_count = 0
+
+        #GC.gc()
+
```
Comment (staticfloat) [inline software/scripts/test_pattern.jl:80]:
```suggestion

```

File: software/scripts/test_pattern.jl [hunk 3]
```diff
@@ -137,31 +151,56 @@ function dma_test(dev_args;use_gpu=false, lfsr_mode=false)
...
 end
 
 function main()
     for dev_args in Devices(driver="XTRX")
+        GC.enable(false)
         try
             dma_test(dev_args; use_gpu=false, lfsr_mode=true)
+            #GC.gc()
             dma_test(dev_args; use_gpu=false, lfsr_mode=false)
+            #GC.gc()
```
Comment (staticfloat) [inline software/scripts/test_pattern.jl:203]:
```suggestion
```

File: software/soapysdr-xtrx/Streaming.cpp [hunk 4]
```diff
@@ -247,7 +250,7 @@ int SoapyXTRX::acquireReadBuffer(SoapySDR::Stream *stream, size_t &handle,
 
     // detect overflows of the underlying circular buffer
     if ((_rx_stream.hw_count - _rx_stream.sw_count) >
-        ((int64_t)_dma_mmap_info.dma_rx_buf_count / 2)) {
+        ((int64_t)_dma_mmap_info.dma_rx_buf_count / 2))  {
```
Comment (staticfloat) [inline software/soapysdr-xtrx/Streaming.cpp:253]:
```suggestion
        ((int64_t)_dma_mmap_info.dma_rx_buf_count / 2)) {
```

**Review by staticfloat — CHANGES_REQUESTED** [review]
The C code changes are easily landed; the julia code changes are a little less clear to me, I think the basic idea is to integrate `TimerOutputs` into the loop, which is a neat idea, can you split your C code changes out into a separate PR, so we can merge those easily, then work on the Julia code by itself?

File: software/scripts/test_pattern.jl [hunk 5]
```diff
@@ -69,19 +73,27 @@ function dma_test(dev_args;use_gpu=false, lfsr_mode=false)
...
-                    initialized_count = false
                     continue
                 elseif err == SoapySDR.SOAPY_SDR_TIMEOUT
                     continue
                 end
 
-                if use_gpu
+                #if i%32 == 0
+                    #println(unsafe_string(SoapySDR.SoapySDRDevice_readSetting(dev, "DMA_BUFFERS")))
+                #end
```
[the remark this answers is not in the document]
Comment (sjkelly) [reply software/scripts/test_pattern.jl:91]: This is handy to have a reference for how to do it for diagnostics. Even if dead code.

File: software/scripts/test_pattern.jl [hunk 6]
```diff
@@ -137,31 +151,56 @@ function dma_test(dev_args;use_gpu=false, lfsr_mode=false)
 
                         # check the data
                         # XXX: this loop does not stay within the 60us time budget
-                        for j in eachindex(buf)
+                        trust_first = false
```
[the remark this answers is not in the document]
Comment (sjkelly) [reply software/scripts/test_pattern.jl:150]: It only checks the first element is sync with the counter. Will add comment...

**Review by staticfloat — APPROVED** [review]

**Pull request merged**